plugin-server/src/cdp/utils/hog-function-filtering.ts

```diff
@@ -0,0 +1,132 @@
+    enabledTelemetry?: boolean
+    /** The event UUID to use for logging */
+    eventUuid?: string
+}): HogFunctionFilterResult {
+    const { hogFunction, filterGlobals, enabledTelemetry, eventUuid } = options
+    const filters = options.filters ?? hogFunction.filters
+    const start = performance.now()
+    const logs: HogFunctionInvocationLogEntry[] = []
+    const metrics: HogFunctionAppMetric[] = []
+
+    let execResult: ExecResult | undefined
+    const result: HogFunctionFilterResult = {
+        match: false,
+        logs,
+        metrics,
+    }
+
+    if (!filters?.bytecode) {
+        result.error = 'No filters bytecode'
+        return result
+    }
+
+    try {
+        execResult = execHog(filters.bytecode, {
+            globals: filterGlobals,
+            telemetry: enabledTelemetry,
+        })
+
+        if (execResult.error) {
+            throw execResult.error
+        }
+
+        result.match = typeof execResult.result === 'boolean' && execResult.result
+
+        if (!result.match) {
+            metrics.push({
+                team_id: hogFunction.team_id,
+                app_source_id: hogFunction.id,
+                metric_kind: 'other',
+                metric_name: 'filtered',
+                count: 1,
+            })
+        }
+    } catch (error) {
+        logger.error('🦔', `[HogFunction] Error filtering function`, {
+            hogFunctionId: hogFunction.id,
+            hogFunctionName: hogFunction.name,
+            teamId: hogFunction.team_id,
+            error: error.message,
+            result: execResult,
+        })
+
+        metrics.push({
+            team_id: hogFunction.team_id,
+            app_source_id: hogFunction.id,
+            metric_kind: 'other',
+            metric_name: 'filtering_failed',
+            count: 1,
+        })
+
+        if (eventUuid) {
+            logs.push({
+                team_id: hogFunction.team_id,
+                log_source: 'hog_function',
+                log_source_id: hogFunction.id,
+                instance_id: new UUIDT().toString(),
+                timestamp: DateTime.now(),
+                level: 'error',
+                message: `Error filtering event ${eventUuid}: ${error.message}`,
+            })
+        }
+        result.error = error.message
+    } finally {
+        const duration = performance.now() - start
+
+        // Re-using the constant from hog-executor.service.ts
+        const DEFAULT_TIMEOUT_MS = 100
+
+        hogFunctionFilterDuration.observe({ type: hogFunction.type }, duration)
+
+        if (duration > DEFAULT_TIMEOUT_MS) {
+            logger.error('🦔', `[HogFunction] Filter took longer than expected`, {
+                hogFunctionId: hogFunction.id,
+                hogFunctionName: hogFunction.name,
+                teamId: hogFunction.team_id,
+                duration,
+                eventId: options?.eventUuid,
+            })
+        }
+    }
+
+    return result
```

Comment:
Generally re-worked this into a single return as it was hard to read the different paths and there was a lot of duplicated logic.